File: Python/2024/2407/240704/BAEK_11725.py
from collections import deque
import sys
input = sys.stdin.readline

N = int(input())
arr = [[] for _ in range(N+1)]
for _ in range(N-1):
    n1, n2 = map(int, input().split())
    arr[n1].append(n2)
    arr[n2].append(n1)

visited = [-1]*(N+1)
queue = deque()
queue.append(1)
visited[1] = 0
while queue:
    now = queue.popleft()

    for next in arr[now]:
        if visited[next] == -1:
            visited[next] = now
            queue.append(next)

print('\n'.join(map(str, visited[2:])))

File: JavaScript/2407/240714/BAEK_11725.js
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin
});

let N;
let graph = [];
let parent = [];
let idx = 0;

const dfs = (node, prev) => {
  parent[node] = prev;

  for (const next of graph[node]) {
    if (next === prev) continue;
    dfs(next, node);
  }
}

rl.on('line', (line) => {
  if (idx === 0) {
    N = parseInt(line);
    graph = Array.from({ length: N+1 }, () => []);
    parent = Array.from({ length: N+1 }, () => 0);
  } else {
    const [a, b] = line.split(' ').map(Number);
    graph[a].push(b);
    graph[b].push(a);
  }

  idx++;
  if (idx === N) {
    rl.close();
  }

}).on('close', () => {
  dfs(1, 0);

  console.log(parent.slice(2).join('\n'));
});